File: source/jugglinglab/prop/ImageProp.java
// ImageProp.java
//
// Copyright 2002-2022 Jack Boyce and the Juggling Lab contributors

package jugglinglab.prop;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.net.URL;
import java.net.MalformedURLException;
import java.io.IOException;
import java.text.MessageFormat;
import javax.imageio.ImageIO;

import jugglinglab.util.*;


public class ImageProp extends Prop {
    static URL image_url_default;
    static {
        image_url_default = ImageProp.class.getResource("/ball.png");
    }

    protected URL url;
    protected BufferedImage image;
    protected BufferedImage scaled_image;
    protected final double width_def = 10.0f;  // in centimeters
    protected double width;
    protected double height;
    protected Dimension size;
    protected Dimension center;
    protected Dimension grip;

    private double last_zoom;


    public ImageProp() throws JuggleExceptionUser {
        if (image_url_default == null)
            throw new JuggleExceptionUser("ImageProp error: Default image not set");
        url = image_url_default;
        loadImage();
        rescaleImage(1.0);
    }

    private void loadImage() throws JuggleExceptionUser {
        try {
            MediaTracker mt = new MediaTracker(new Component() {});
            image = ImageIO.read(url);
            mt.addImage(image, 0);
            // Try to laod the image
            try {
                mt.waitForAll();
            } catch (InterruptedException ex) {}

            if (mt.isErrorAny()) {
                image = null;
                // This could also be bad image data, but it is usually a nonexistent file.
                throw new JuggleExceptionUser(errorstrings.getString("Error_bad_file"));
            }

            double aspectRatio = ((double)image.getHeight()) / ((double)image.getWidth());
            width = width_def;
            height = width_def * aspectRatio;
        } catch (IOException e) {
            throw new JuggleExceptionUser(errorstrings.getString("Error_bad_file"));
        } catch (SecurityException se) {
            throw new JuggleExceptionUser(errorstrings.getString("Error_security_restriction"));
        }
    }

    private void rescaleImage(double zoom) {
        int image_pixel_width = (int)(0.5 + zoom * width);
        if (image_pixel_width < 1)
            image_pixel_width = 1;
        int image_pixel_height = (int)(0.5 + zoom * height);
        if (image_pixel_height < 1)
            image_pixel_height = 1;
        size = new Dimension(image_pixel_width, image_pixel_height);
        center = new Dimension(image_pixel_width/2, image_pixel_height/2);

        int offsetx = image_pixel_width / 2;
        int offsety = image_pixel_height;
        grip = new Dimension(offsetx, offsety);

        last_zoom = zoom;

        scaled_image = new BufferedImage(image_pixel_width, image_pixel_height, image.getType());
        Graphics2D g = scaled_image.createGraphics();
        g.setRenderingHint(RenderingHints.KEY_INTERPOLATION,
                           RenderingHints.VALUE_INTERPOLATION_BILINEAR);
        g.drawImage(image, 0, 0, image_pixel_width, image_pixel_height,
                    0, 0, image.getWidth(), image.getHeight(), null);
        g.dispose();
    }

    // View methods

    @Override
    public String getType() {
        return "Image";
    }

    @Override
    public Color getEditorColor() {
        // The color that shows up in the visual editor
        // We could try to get an average color for the image
        return Color.white;
    }

    @Override
    public ParameterDescriptor[] getParameterDescriptors() {
        ParameterDescriptor[] result = new ParameterDescriptor[2];

        result[0] = new ParameterDescriptor("image", ParameterDescriptor.TYPE_ICON,
                            null, image_url_default, url);
        result[1] = new ParameterDescriptor("width", ParameterDescriptor.TYPE_FLOAT,
                            null, Double.valueOf(width_def), Double.valueOf(width));

        return result;
    }

    @Override
    protected void init(String st) throws JuggleExceptionUser {
        if (st == null) return;

        ParameterList pl = new ParameterList(st);

        String sourcestr = pl.getParameter("image");
        if (sourcestr != null) {
            try {
                url = new URL(sourcestr);
                loadImage();
            } catch (MalformedURLException ex) {
                throw new JuggleExceptionUser(errorstrings.getString("Error_malformed_URL"));
            }
        }

        String widthstr = pl.getParameter("width");
        if (widthstr != null) {
            try {
                double temp = Double.valueOf(widthstr).doubleValue();
                if (temp > 0) {
                    width = temp;
                    double aspectRatio = ((double)image.getHeight(null)) / ((double)image.getWidth(null));
                    height = width * aspectRatio;
                }
                else
                    throw new NumberFormatException();
            } catch (NumberFormatException nfe) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "width" };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }
    }

    @Override
    public Image getProp2DImage(double zoom, double[] camangle) {
        if (zoom != last_zoom)
            rescaleImage(zoom);
        return scaled_image;
    }

    @Override
    public Coordinate getMax() {
        return new Coordinate(width / 2.0, 0.0, width);
    }

    @Override
    public Coordinate getMin() {
        return new Coordinate(-width / 2.0, 0.0, 0.0);
    }

    @Override
    public double getWidth() {
        return width;
    }

    @Override
    public Dimension getProp2DSize(double zoom) {
        if (size == null || zoom != last_zoom)
            rescaleImage(zoom);
        return size;
    }

    @Override
    public Dimension getProp2DCenter(double zoom) {
        if (center == null || zoom != last_zoom)
            rescaleImage(zoom);
        return center;
    }

    @Override
    public Dimension getProp2DGrip(double zoom) {
        if (grip == null || zoom != last_zoom)
            rescaleImage(zoom);
        return grip;
    }
}

File: src/main/kotlin/jugglinglab/prop/ImageProp.kt
//
// ImageProp.kt
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.prop

import jugglinglab.JugglingLab.errorstrings
import jugglinglab.util.Coordinate
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.ParameterDescriptor
import jugglinglab.util.ParameterList
import jugglinglab.util.jlParseFiniteDouble
import java.awt.*
import java.awt.image.BufferedImage
import java.io.IOException
import java.net.MalformedURLException
import java.net.URI
import java.net.URISyntaxException
import java.net.URL
import java.text.MessageFormat
import javax.imageio.ImageIO
import kotlin.math.max

class ImageProp : Prop() {
    private var url: URL?
    private var image: BufferedImage? = null
    private var scaledImage: BufferedImage? = null
    private var width: Double = 0.0
    private var height: Double = 0.0
    private var size: Dimension? = null
    private var center: Dimension? = null
    private var grip: Dimension? = null
    private var lastZoom = 0.0

    init {
        url = imageUrlDefault ?:
            throw JuggleExceptionUser("ImageProp error: Default image not set")
        loadImage()
        rescaleImage(1.0)
    }

    @Throws(JuggleExceptionUser::class)
    private fun loadImage() {
        try {
            val mt = MediaTracker(object : Component() {})
            image = ImageIO.read(url)
            mt.addImage(image, 0)
            // Try to laod the image
            try {
                mt.waitForAll()
            } catch (_: InterruptedException) {
            }

            if (mt.isErrorAny()) {
                image = null
                // This could also be bad image data, but it is usually a nonexistent file.
                throw JuggleExceptionUser(errorstrings.getString("Error_bad_file"))
            }

            val aspectRatio = (image!!.height.toDouble()) / (image!!.width.toDouble())
            width = WIDTH_DEF
            height = WIDTH_DEF * aspectRatio
        } catch (_: IOException) {
            throw JuggleExceptionUser(errorstrings.getString("Error_bad_file"))
        } catch (_: SecurityException) {
            throw JuggleExceptionUser(errorstrings.getString("Error_security_restriction"))
        }
    }

    private fun rescaleImage(zoom: Double) {
        val imagePixelWidth = max((0.5 + zoom * width).toInt(), 1)
        val imagePixelHeight = max((0.5 + zoom * height).toInt(), 1)
        size = Dimension(imagePixelWidth, imagePixelHeight)
        center = Dimension(imagePixelWidth / 2, imagePixelHeight / 2)

        val offsetx = imagePixelWidth / 2
        val offsety = imagePixelHeight
        grip = Dimension(offsetx, offsety)

        lastZoom = zoom

        scaledImage = BufferedImage(imagePixelWidth, imagePixelHeight, image!!.type)
        val g = scaledImage!!.createGraphics()
        g.setRenderingHint(
            RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BILINEAR
        )
        g.drawImage(
            image,
            0,
            0,
            imagePixelWidth,
            imagePixelHeight,
            0,
            0,
            image!!.width,
            image!!.height,
            null
        )
        g.dispose()
    }

    override val type = "Image"

    override val isColorable = false

    override fun getEditorColor(): Color {
        // The color that shows up in the visual editor
        // We could try to get an average color for the image
        return Color.white
    }

    override fun getParameterDescriptors(): Array<ParameterDescriptor> {
        return arrayOf(
            ParameterDescriptor(
                "image",
                ParameterDescriptor.TYPE_ICON,
                null,
                imageUrlDefault,
                url
            ),
            ParameterDescriptor(
                "width",
                ParameterDescriptor.TYPE_FLOAT,
                null,
                WIDTH_DEF,
                width
            ),
        )
    }

    @Throws(JuggleExceptionUser::class)
    override fun init(st: String?) {
        if (st == null) {
            return
        }
        val pl = ParameterList(st)

        val sourcestr = pl.getParameter("image")
        if (sourcestr != null) {
            try {
                url = URI(sourcestr).toURL()
                loadImage()
            } catch (_: URISyntaxException) {
                throw JuggleExceptionUser(errorstrings.getString("Error_malformed_URL"))
            } catch (_: MalformedURLException) {
                throw JuggleExceptionUser(errorstrings.getString("Error_malformed_URL"))
            }
        }

        val widthstr = pl.getParameter("width")
        if (widthstr != null) {
            try {
                val temp = jlParseFiniteDouble(widthstr)
                if (temp > 0) {
                    width = temp
                    val aspectRatio = (image!!.getHeight(null).toDouble()) / (image!!.getWidth(null).toDouble())
                    height = width * aspectRatio
                } else {
                    throw NumberFormatException()
                }
            } catch (_: NumberFormatException) {
                val template: String = errorstrings.getString("Error_number_format")
                val arguments = arrayOf<Any?>("width")
                throw JuggleExceptionUser(MessageFormat.format(template, *arguments))
            }
        }
    }

    override fun getProp2DImage(zoom: Double, camangle: DoubleArray): Image? {
        if (zoom != lastZoom) {
            rescaleImage(zoom)
        }
        return scaledImage
    }

    override fun getMax(): Coordinate {
        return Coordinate(width / 2, 0.0, width)
    }

    override fun getMin(): Coordinate {
        return Coordinate(-width / 2, 0.0, 0.0)
    }

    override fun getWidth(): Double {
        return width
    }

    override fun getProp2DSize(zoom: Double): Dimension? {
        if (size == null || zoom != lastZoom) {
            rescaleImage(zoom)
        }
        return size
    }

    override fun getProp2DCenter(zoom: Double): Dimension? {
        if (center == null || zoom != lastZoom) {
            rescaleImage(zoom)
        }
        return center
    }

    override fun getProp2DGrip(zoom: Double): Dimension? {
        if (grip == null || zoom != lastZoom) {
            rescaleImage(zoom)
        }
        return grip
    }

    companion object {
        private const val WIDTH_DEF: Double = 10.0  // in centimeters
        private var imageUrlDefault: URL? = ImageProp::class.java.getResource("/ball.png")
    }
}
